### ISSUE TITLE
Unstable cilium-health status after update to 1.11.0

### ISSUE THREAD

**Issue by IMMORTALxJO** [cilium/cilium#18302]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

I have upgraded test cluster to 1.11.0 ( from 1.10.6 ) and got unreachable endpoints in `cilium-health status` on some nodes. Looks like health IP addresses are not synchronized between cluster nodes.
Example cilium-health output for one of unstable nodes - **ya-test-k8s-worker-02**:
```
#### pod: cilium-86brf   2/2   Running   2     55m   10.128.0.50   ya-test-k8s-master-01   <none>   <none>
  ya-test-k8s/ya-test-k8s-worker-02:
    Host connectivity to 10.128.0.55:
      ICMP to stack:   OK, RTT=1.294383ms
      HTTP to agent:   OK, RTT=931.421µs
    Endpoint connectivity to 10.10.4.184:
      ICMP to stack:   OK, RTT=991.167µs
      HTTP to agent:   Get "http://10.10.4.184:4240/hello": dial tcp 10.10.4.184:4240: connect: connection timed out
#### pod: cilium-bhckj   2/2   Running   2     55m   10.128.0.42   ya-test-k8s-master-02   <none>   <none>
  ya-test-k8s/ya-test-k8s-worker-02:
    Host connectivity to 10.128.0.55:
      ICMP to stack:   OK, RTT=1.364705ms
      HTTP to agent:   OK, RTT=646.87µs
    Endpoint connectivity to 10.10.4.139:
      ICMP to stack:   OK, RTT=1.517272ms
      HTTP to agent:   OK, RTT=766.268µs
#### pod: cilium-fct6h   2/2   Running   2     55m   10.128.0.10   ya-test-k8s-master-03   <none>   <none>
  ya-test-k8s/ya-test-k8s-worker-02:
    Host connectivity to 10.128.0.55:
      ICMP to stack:   OK, RTT=1.791616ms
      HTTP to agent:   OK, RTT=903.631µs
    Endpoint connectivity to 10.10.4.184:
      ICMP to stack:   OK, RTT=885.139µs
      HTTP to agent:   Get "http://10.10.4.184:4240/hello": dial tcp 10.10.4.184:4240: connect: connection timed out
#### pod: cilium-nmgjc   2/2   Running   2     55m   10.128.0.21   ya-test-k8s-worker-01   <none>   <none>
  ya-test-k8s/ya-test-k8s-worker-02:
    Host connectivity to 10.128.0.55:
      ICMP to stack:   OK, RTT=3.704711ms
      HTTP to agent:   OK, RTT=1.409363ms
    Endpoint connectivity to 10.10.4.139:
      ICMP to stack:   OK, RTT=1.287935ms
      HTTP to agent:   OK, RTT=1.085519ms
#### pod: cilium-t9kjx   2/2   Running   2     55m   10.128.0.55   ya-test-k8s-worker-02   <none>   <none>
  ya-test-k8s/ya-test-k8s-worker-02 (localhost):
    Host connectivity to 10.128.0.55:
      ICMP to stack:   OK, RTT=805.184µs
      HTTP to agent:   OK, RTT=259.144µs
    Endpoint connectivity to 10.10.4.139:
      ICMP to stack:   OK, RTT=790.565µs
      HTTP to agent:   OK, RTT=573.987µs
```
2 of 5 nodes fail endpoint health check due to the wrong health endpoint IP address ( 10.10.4.184 instead of 10.10.4.139 ). Ciliumnode resource has correct ip address:
```
# kubectl get ciliumnodes ya-test-k8s-worker-02 -o json | jq -r '.spec.health.ipv4'
10.10.4.139
```
also this IP address can be found in bpf ipcache and in the cilium map of a "broken" agents:
```
# kubectl exec -it -n kube-system cilium-86brf -c cilium-agent -- cilium map get cilium_ipcache | grep '10.10.2.237'
10.10.2.237/32   4 0 10.128.0.42       sync
# kubectl exec -it -n kube-system cilium-86brf -c cilium-agent -- cilium bpf ipcache list | grep '10.10.2.237'
10.10.2.237/32      4 0 10.128.0.42
```
Already tried to restart cilium-agent pods, waited for an hours but endpoints are still unhealthy.

### Cilium Version

1.11.0

### Kernel Version

4.19.0-0.bpo.17-amd64 #1 SMP Debian 4.19.194-3~deb9u1 (2021-07-19) x86_64 GNU/Linux

### Kubernetes Version

v1.21.5

### Sysdump

[cilium-sysdump-20211218-160750.zip](https://disk.yandex.ru/d/TvjRV8I07U9oow)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by ghouscht** [comment]
We're affected by the same issue. Tracked this a bit down and the last "good" version seems to be 1.11-rc1. I see the behavior described in this issue in all versions from 1.11-rc2 on. Will dig a bit deeper and see if I can find something more...

**Comment by aanm** [comment]
@IMMORTALxJO thank you for the GH issue, can you reproduce the issue with debug logs enabled? Set `debug: true` in the Cilium's configmap. Thank you.

**Comment by ghouscht** [comment]
> @IMMORTALxJO thank you for the GH issue.
> 
>     1. Can you reproduce the issue with debug logs enabled? Set `debug: true` in the Cilium's configmap.
> 
>     2. Can you provide the output of `cilium node list -o json` from one of the nodes that has its connectivity failing?
> 
> 
> Thank you.

1. Yes, can send you the logs or an entire sysdump to your e-mail, sharing publicly is not possible, sorry!
2. Output of `cilium-health status` on one of the nodes with failing connectivity:
```plaintext
Probe time:   2022-01-05T09:52:50Z
Nodes:
  e1-k8s-pfnet-lab-b/e1-k8s-mlsu105.pnet.ch (localhost):
    Host connectivity to 172.16.177.197:
      ICMP to stack:   OK, RTT=487.894µs
      HTTP to agent:   OK, RTT=169.162µs
    Endpoint connectivity to 10.127.65.45:
      ICMP to stack:   OK, RTT=451.547µs
      HTTP to agent:   OK, RTT=239.062µs
  e1-k8s-pfnet-lab-b/e1-k8s-alsu101.pnet.ch:
    Host connectivity to 172.16.176.123:
      ICMP to stack:   OK, RTT=1.019811ms
      HTTP to agent:   OK, RTT=1.157778ms
    Endpoint connectivity to 10.127.66.29:
      ICMP to stack:   OK, RTT=1.483466ms
      HTTP to agent:   Get "http://10.127.66.29:4240/hello": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
...
``` 
Output of  `k get ciliumnode e1-k8s-alsu101.pnet.ch -o json | jq '.spec.health'`:
```json
{
  "ipv4": "10.127.66.98"
}
```
Output of  `cilium node list -o json`:
``` json
[
  {
    "health-endpoint-address": {
      "ipv4": {
        "enabled": true,
        "ip": "10.127.65.151"
      },
      "ipv6": {}
    },
    "name": "e1-k8s-pfnet-lab-b/e1-k8s-alsu102.pnet.ch",
    "primary-address": {
      "ipv4": {
        "address-type": "InternalIP",
        "alloc-range": "10.127.65.128/25",
        "enabled": true,
        "ip": "172.16.176.83"
      },
      "ipv6": {}
    },
    "secondary-addresses": [
      {
        "address-type": "CiliumInternalIP",
        "ip": "10.127.65.204"
      }
    ]
  },
  {
    "health-endpoint-address": {
      "ipv4": {
        "enabled": true,
        "ip": "10.127.66.235"
      },
      "ipv6": {}
    },
    "name": "e1-k8s-pfnet-lab-b/e1-k8s-alsu103.pnet.ch",
    "primary-address": {
      "ipv4": {
        "address-type": "InternalIP",
        "alloc-range": "10.127.66.128/25",
        "enabled": true,
        "ip": "172.16.176.124"
      },
      "ipv6": {}
    },
    "secondary-addresses": [
      {
        "address-type": "CiliumInternalIP",
        "ip": "10.127.66.186"
      }
    ]
  },
  {
    "health-endpoint-address": {
      "ipv4": {
        "enabled": true,
        "ip": "10.127.64.106"
      },
      "ipv6": {}
    },
    "name": "e1-k8s-pfnet-lab-b/e1-k8s-mlsu104.pnet.ch",
    "primary-address": {
      "ipv4": {
        "address-type": "InternalIP",
        "alloc-range": "10.127.64.0/25",
        "enabled": true,
        "ip": "172.16.177.200"
      },
      "ipv6": {}
    },
    "secondary-addresses": [
      {
        "address-type": "CiliumInternalIP",
        "ip": "10.127.64.109"
      }
    ]
  },
  {
    "health-endpoint-address": {
      "ipv4": {
        "enabled": true,
        "ip": "10.127.65.45"
      },
      "ipv6": {}
    },
    "name": "e1-k8s-pfnet-lab-b/e1-k8s-mlsu105.pnet.ch",
    "primary-address": {
      "ipv4": {
        "address-type": "InternalIP",
        "alloc-range": "10.127.65.0/25",
        "enabled": true,
        "ip": "172.16.177.197"
      },
      "ipv6": {}
    },
    "secondary-addresses": [
      {
        "address-type": "CiliumInternalIP",
        "ip": "10.127.65.38"
      }
    ]
  },
  {
    "health-endpoint-address": {
      "ipv4": {
        "enabled": true,
        "ip": "10.127.64.157"
      },
      "ipv6": {}
    },
    "name": "e1-k8s-pfnet-lab-b/e1-k8s-mlsu106.pnet.ch",
    "primary-address": {
      "ipv4": {
        "address-type": "InternalIP",
        "alloc-range": "10.127.64.128/25",
        "enabled": true,
        "ip": "172.16.177.196"
      },
      "ipv6": {}
    },
    "secondary-addresses": [
      {
        "address-type": "CiliumInternalIP",
        "ip": "10.127.64.167"
      }
    ]
  },
  {
    "health-endpoint-address": {
      "ipv4": {
        "enabled": true,
        "ip": "10.127.66.98"
      },
      "ipv6": {}
    },
    "name": "e1-k8s-pfnet-lab-b/e1-k8s-alsu101.pnet.ch",
    "primary-address": {
      "ipv4": {
        "address-type": "InternalIP",
        "alloc-range": "10.127.66.0/25",
        "enabled": true,
        "ip": "172.16.176.123"
      },
      "ipv6": {}
    },
    "secondary-addresses": [
      {
        "address-type": "CiliumInternalIP",
        "ip": "10.127.66.78"
      }
    ]
  }
]
```

The node with failing connectivity (mls105) still thinks the health-ip of alsu101 is 10.127.66.29 but as you can see in the output above it is 10.127.66.**98**.

**Comment by aanm** [comment]
After adding this line https://github.com/cilium/cilium/pull/18382/files#diff-e9cec4c2460e089f0f518c07f3d4c2fb40273486a8346d499ec5188395417772R445 and reproduce it again, I've seen this in the logs:
```
2022-01-05T15:22:08.075269839+01:00 level=debug msg="ipcache IP upsert error" error="unable to overwrite source \"kube-apiserver\" with source \"custom-resource\"" ip-address="{InternalIP 172.16.177.197}" subsys=nodemanager
2022-01-05T15:22:08.075320081+01:00 level=debug msg="ipcache IP upsert error" error="unable to overwrite source \"kube-apiserver\" with source \"custom-resource\"" ip-address="{CiliumInternalIP 10.127.65.38}" subsys=nodemanager
```

This prevents the cilium health address from being updated in the list of cilium nodes because `dpUpdate` is false and the branch https://github.com/cilium/cilium/pull/18382/files#diff-e9cec4c2460e089f0f518c07f3d4c2fb40273486a8346d499ec5188395417772R488 is never executed.

cc @christarazi

**Comment by christarazi** [comment]
I believe this will be fixed by #18150. It would be great if anyone affected by this bug can try out that PR.

**Comment by ghouscht** [comment]
> I believe this will be fixed by #18150. It would be great if anyone affected by this bug can try out that PR.

I tried but the images seem to be gone (maybe expired?). I tried the images found [in here](https://github.com/cilium/cilium/runs/4600951710?check_suite_focus=true)

**Comment by ghouscht** [comment]
> > I believe this will be fixed by #18150. It would be great if anyone affected by this bug can try out that PR.
> 
> I tried but the images seem to be gone (maybe expired?). I tried the images found [in here](https://github.com/cilium/cilium/runs/4600951710?check_suite_focus=true)

I'm no longer able to reproduce the issue with an image from https://github.com/cilium/cilium/pull/18390. Tested with the following image:
```
quay.io/cilium/cilium-ci:57e71d5aa8acdc03fddc041b8fee9b13f054858a@sha256:43c2807ef4e27e28140ee7072a9173be25fb825d8d0b4b0dcef1adc16e5bd0ae
```

**Comment by christarazi** [comment]
Given that we have confirmation that https://github.com/cilium/cilium/pull/18150 fixes this issue (https://github.com/cilium/cilium/pull/18390 for v1.11 backport), closing.
